rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -0,0 +1,724 @@
+///
+/// This operation is similar to SQL's MERGE statement. It allows you to merge
+/// new data with existing data.
+///
+/// Use the [MergeInsertBuilder] to construct an merge insert job. For example:
+///
+/// ```ignore
+/// // find-or-create, insert new rows only
+/// let dataset = MergeInsertBuilder::new(dataset, vec!["my_key"])
+///     .build()?
+///     .execute(new_data)
+///     .await?;
+///
+/// // upsert, insert or update
+/// let dataset = MergeInsertBuilder::new(dataset, vec!["my_key"])
+///     .when_not_matched(WhenNotMatched::UpdateAll)
+///     .build()?
+///     .execute(new_data)
+///     .await?;
+///
+/// // replace data for month=january
+/// let dataset = MergeInsertBuilder::new(dataset, vec!["my_key"])
+///     .when_not_matched(WhenNotMatched::UpdateAll)
+///     .when_not_matched_by_source(
+///         WhenNotMatchedBySource::DeleteIf(month_eq_jan)
+///     )
+///     .build()?
+///     .execute(new_data)
+///     .await?;
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct MergeInsertBuilder {
+    dataset: Arc<Dataset>,
+    params: MergeInsertParams,
+}
+
+impl MergeInsertBuilder {
+    /// Creates a new builder
+    ///
+    /// By default this will build a job that has the same semantics as find-or-create
+    ///  - matching rows will be kept as-is
+    ///  - new rows in the new data will be inserted
+    ///  - rows in the old data that do not match will be left as-is
+    ///
+    /// Use the methods on this builder to customize that behavior
+    pub fn new(dataset: Arc<Dataset>, on: Vec<String>) -> Self {
+        Self {
+            dataset,
+            params: MergeInsertParams {
+                on,
+                update_matched: false,
+                insert_not_matched: true,
+                delete_not_matched_by_source: WhenNotMatchedBySource::NeverDelete,
+            },
+        }
+    }
+
+    /// Specify what should happen when a target row matches a row in the source
+    pub fn when_matched(mut self, behavior: WhenMatched) -> Self {
+        self.params.update_matched = match behavior {
+            WhenMatched::DoNothing => false,
+            WhenMatched::UpdateAll => true,
+        };
+        self
+    }
+
+    /// Specify what should happen when a source row has no match in the target
+    ///
+    /// These are typically "new rows"
+    pub fn when_not_matched(mut self, behavior: WhenNotMatched) -> Self {
+        self.params.insert_not_matched = match behavior {
+            WhenNotMatched::DoNothing => false,
+            WhenNotMatched::InsertAll => true,
+        };
+        self
+    }
+
+    /// Specify what should happen when a target row has no match in the source
+    ///
+    /// These are typically "old rows"
+    pub fn when_not_matched_by_source(mut self, behavior: WhenNotMatchedBySource) -> Self {
+        self.params.delete_not_matched_by_source = behavior;
+        self
+    }
+
+    /// Crate a merge insert job
+    pub fn build(&mut self) -> MergeInsertJob {
```

Comment:
Should we validate here that `!self.on.is_empty()`?

Reply:
I changed `MergeInsertJob::new` to `MergeInsertJob::try_new` and validate there.